ze/btx_zeinterval_callbacks.cpp

```diff
@@ -765,87 +765,281 @@ static void zeEventDestroy_exit_callback(void *btx_handle, void *usr_data, int64
+    double wtBandwidth =
+        static_cast<double>(pMemBandwidth_val->writeCounter - prev_bandwidth.writeCounter) * 1e6 /
+        (time_diff);
+    DeviceHash uuid_idx = get_device_hash(usr_data, hostname, vpid, hDevice);
+    btx_push_message_sampling_memModule(btx_handle, hostname, prev_ts, uuid_idx.hash,
+                                        uuid_idx.deviceIdx, (uint64_t)hMemModule, subDevice,
+                                        pBandwidth, rdBandwidth, wtBandwidth, allocation);
+    // Update the stored values
+    it->second = {*pMemBandwidth_val, ts};
+  } else {
+    std::cerr << "Memory property not found!" << std::endl;
+  }
+}
 
-  btx_push_message_lttng_frequency(btx_handle, hostname, 0, 0, ts, BACKEND_ZE, (uint64_t)hDevice,
-                                   domain, frequency);
+static void lttng_ust_ze_sampling_engineStats_callback(void *btx_handle, void *usr_data, int64_t ts,
+                                                       const char *hostname, int64_t vpid,
+                                                       uint64_t vtid, ze_device_handle_t hDevice,
+                                                       zes_engine_handle_t hEngine,
+                                                       size_t _pEngineStats_val_length,
+                                                       zes_engine_stats_t *pEngineStats_val) {
+  auto *data = static_cast<data_t *>(usr_data);
+  const auto it0 = data->engine_property.find({hostname, vpid, hDevice, hEngine});
+  if (it0 != data->engine_property.cend()) {
+    const auto &engineProps = it0->second;
+    uint32_t subDevice = engineProps.subdeviceId;
+
+    if (engineProps.type == ZES_ENGINE_GROUP_COMPUTE_ALL ||
+        engineProps.type == ZES_ENGINE_GROUP_COPY_ALL) {
+      auto [it, inserted] = data->device_engines_ref.insert(
+          {{hostname, vpid, hDevice, hEngine, subDevice}, {*pEngineStats_val, ts}});
+      if (inserted)
+        return;
+      auto &[prev_engineStats, prev_ts] = it->second;
+      if (pEngineStats_val->timestamp == prev_engineStats.timestamp)
+        return;
+      double time_diff =
+          pEngineStats_val->timestamp >= prev_engineStats.timestamp
+              ? static_cast<double>(pEngineStats_val->timestamp - prev_engineStats.timestamp)
+              : static_cast<double>(pEngineStats_val->timestamp +
+                                    (UINT64_MAX - prev_engineStats.timestamp) + 1);
+      double activeTime = 0;
+      if (pEngineStats_val->activeTime > prev_engineStats.activeTime)
+        activeTime =
+            static_cast<double>(pEngineStats_val->activeTime - prev_engineStats.activeTime) * 100 /
+            time_diff;
+      else
+        activeTime = static_cast<double>((UINT64_MAX - prev_engineStats.activeTime) +
+                                         pEngineStats_val->activeTime + 1) *
+                     100 / time_diff;
+      DeviceHash uuid_idx = get_device_hash(usr_data, hostname, vpid, hDevice);
```

Comment:
You don't test tht this guy doesn't return `{0,0}` (or optional not set)
Do you want to crash or to not send an message in this case?

Reply:
I used optional and a message if null value returned.